### DISCUSSION TITLE
Block templates service and REST API

### DISCUSSION THREAD

**Discussion by joshuatf** [woocommerce/woocommerce#39740] [WooCommerce New Product Editor] [answered]
As we migrate the product block editor to the newly created extensible `BlockTemplate` classes, we'll need to consider both how these are exposed to developers to allow them to be extended as well as how we can retrieve them via REST API.

## Exposing the templates

For exposing the templates, we'll need a registry to keep track of registered block templates.  These templates at a minimum need a few distinguishing features to help identify their use:

* A unique ID
* A category (or area) in which the template is predominantly used
* A method to get the prepared template for use in the client

It's worth noting that while the [WP_Block_Template](https://developer.wordpress.org/reference/classes/wp_block_template/) includes some redundant information when it comes to our registered templates, it also includes properties such as ID, slug, area, title, and description which could all prove useful.

At a minimum, the registry should allow getting all templates and getting a template by ID, but some helper methods to get template by area or other properties could be nice to have.

```php
$template_registry = wc_get_container()->get( BlockTemplateRegistry::class )->get_all_registered();
$template_registry = wc_get_container()->get( BlockTemplateRegistry::class )->get_template( 'template-id' );
```

## Serving the templates via REST API

Currently the product editor template is hydrated in the editor settings on page load, but this occurs on all WooCommerce Admin pages.  Hydrating these on every page is not sustainable as the admin Reactified pages grow in size and we'll need to offload this to a REST API.  I see 3 options for this:

### 1. Use the existing templates REST API

WP already has [a REST API for templates](https://github.com/WordPress/wordpress-develop/blob/6.3/src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php).  The only minor issue is that the `content.raw` property in the response is meant to be a comment-delimited template in string format (`<!-- wp:blockName ... ->`).  The API was meant to serve the saved versions of templates from the database.

The templates that are used in the product editor and soon-to-come areas of WooCommerce will use the array format since they are not yet instantiated templates (`[ $blockName, $blockAttrs, $blockChidren ]`).

Using an array here does not coerce this value to a string and does work for our purposes, but obviously doesn't match the expected schema.  But to circumvent that, we could pass the array as a string and decode in the client.

**Advantages**

* The REST API is already in place and little to no work is needed to get this up and running
* Little to no maintenance costs since WP core maintains this endpoint
* Provides a number of tools in the client to handle fetching templates out of the box

**Disadvantages**

* Feels slightly hacky and we need to work around mismatched schema in a couple of places
* To make this work with filters for the endpoint, we can only return templates if the consumer knows a key property to identify the template (`area`, `id`, etc)

### 2. Extend the current templates REST API

Another option is to extend the existing templates API, overriding the parent to include the schema needed for our use case.

```php
class WooCommerceTemplatesController extends WP_REST_Templates_Controller{
    // Override the schema
    // Override the namespace and endpoint
}
```

Short-term, this gives us a slightly more flexible solution with less development time, but ultimately falls short of the other two solutions long-term in my opinion.

**Advantages**

* Slightly more flexible than option 1
* Allows faster development than spinning up our own API from scratch

**Disadvantages**

* Does not immediately provide us with the client utils and data stores for accessing templates
* Could create more maintenance issues long-term since the core endpoint class is not namespaced or versioned

### 3. Spin up our own endpoint for templates

The most flexible solution of course is to spin up our own endpoint for handling these templates, but comes with the highest development cost.

**Advantages**

* Most flexible option
* Reduces concerns around lack of versioning or changes of the core endpoint

**Disadvantages**

* Requires the most upfront development time
* Requires client-side data stores and utils to be developed, unless we create an entity type that the entity data stores can access

## Discussion

I have a https://github.com/woocommerce/woocommerce/pull/39698 that uses option 1 from above.

* Are there other pieces of information that might be useful for templates?  Any type of meta information that's needed to display them on the frontend?
* Which of the REST options looks to be the most viable?  I'm leaning towards option 1 as I feel this accomplishes our goals and requires the least development time and maintenance.  Additionally, I feel this is a feature that should ultimately reside in WP, but few plugins are using templates to their full ability thus far so little thought is being put into non-instantiated templates.
* Are there any cases with the product editor or other admin pages that will be converted in the future that make the current POC not viable?

**Comment by louwie17** [comment]
> Are there other pieces of information that might be useful for templates? Any type of meta information that's needed to display them on the frontend?

The only info I can think off would be a title, description, maybe a icon of some sort, and potentially an plugin id. Incase we will create a separate screen displaying the product types. But looks like title/description are already available, so that may be enough for now.
In terms of Meta, it's hard to say, I could see a case where we did want to retrieve a template derived from specific product data that does not tie to a specific `id`. But maybe this is something we can add on top of the above API as this is more WooCommerce centric. We did definitely need some kind of look up store to get the correct template id.

> Which of the REST options looks to be the most viable? I'm leaning towards option 1 as I feel this accomplishes our goals and requires the least development time and maintenance. Additionally, I feel this is a feature that should ultimately reside in WP, but few plugins are using templates to their full ability thus far so little thought is being put into non-instantiated templates.

I am leaning towards option 1 as well, with option 2 as a close second, incase we want to adjust our data a bit and remove the replace the `content` string format with the array format. But maybe we can already do this by using a filter? In either case it seems valuable to me that the REST api should support providing that option anyhow, allowing us to say how we want the template content (default can be the string format).

> Are there any cases with the product editor or other admin pages that will be converted in the future that make the current POC not viable?

I am not sure.

**Reply by joshuatf** [reply]
>But maybe we can already do this by using a filter? 

The good news is that even with option 1, we can add to the existing field with the `register_rest_field` method so this is an option.

Currently the only other potential field I see is some type of logic for "conditional fields" if we choose to pass a ruleset as opposed to using conditional blocks.  But it remains to be determined which route we opt for there.

**Comment by nerrad** [comment]
> The templates that are used in the product editor and soon-to-come areas of WooCommerce will use the array format since they are not yet instantiated templates ([ $blockName, $blockAttrs, $blockChidren ]).

How is the template parsed for the product editor?

**Reply by joshuatf** [reply]
It's fed as the above [array format](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-templates/):

```php
[ $blockName, $blockAttrs, $blockChidren ]
```

And then synchronized directly with the editor template blocks:

```
synchronizeBlocksWithTemplate( [], _settings?.template ),
```

Given the blocking work on @woocommerce/mothra and lack of opposition, I opted to use the existing templates REST endpoint to serve these templates in https://github.com/woocommerce/woocommerce/pull/39698.

This can easily be swapped out with our REST endpoint at any point in the future, though additional client-side hooks to wrap the entity calls will prevent breaking public API.  E.g., `useProductTemplates()` as opposed to directly grabbing templates via the entity data store.

**Reply by nerrad** [reply]
I see, I'm curious why you don't use the client side parser for the block formatted string via the REST API? Similar to what is done in the `setupEditorStore` here:

https://github.com/WordPress/gutenberg/blob/2d4314d754dcec540ac05890cc0d2dbf89ab5d53/packages/editor/src/store/actions.js#L52

**Reply by joshuatf** [reply]
I _think_ that the string format needs to be rendered first in the client and then persisted into the database or at least have knowledge of the structure of certain blocks as is done in template files.  For example, `<!-- wp:list-item -->` expects the first child element to be a `<li>`.

This does not work particularly well for the template API since we need knowledge of the `save` method in order to correctly format this string.

It does work well for editors that persist templates to the database since they would have been processed by the block's `save` method client-side.

**Comment by AnnaMag** [comment]
>Are there other pieces of information that might be useful for templates? Any type of meta information that's needed to display them on the frontend?

Is it a good time to think about template versioning? Sort of sub-templates thought of as a personalized copy of a template with modification that would be stored for re-use, e.g. create/edit products always from a personalized template.

**Reply by joshuatf** [reply]
This is an interesting idea.  Do you envision merchants creating a customized version of the product edit template in some way?

Currently these templates diverge from the traditional templates in that they don't have any persistence.  If we do store customized copies, this opens up a lot of discussion about whether or not those persisted templates are still modifiable by third parties.

**Reply by AnnaMag** [reply]
Imagine a flow, where users create a product based on a default template X, customize it (e.g. modify fields), and save the product as a template for re-use, X1. Then, when creating a product they are offered an option to start from scratch (from X) or choose to create a product like the existing one(s) (from X1). 
In this sense, every product would have access to the info about its associated template (1-1 correspondence), such that its editing happens in the correct view/template.

> this opens up a lot of discussion about whether or not those persisted templates are still modifiable by third parties.

The templates are site-specific, in use for products on a given site, so modifiable "locally". They could also be thought of as shareable objects, but this might not be the immediate goal. Is this what you meant by "modifiable by third parties"?

**Reply by joshuatf** [reply]
I can see this being especially useful for creating a custom template (not necessarily tied to a specific product type) which simplifies expectations around this template.  I think this is perfectly feasible, though we'll face a few technical hurdles that we'll need to address:

1. Determining where these templates should be stored.  Most likely as a `wp_template` type in the posts table, but we need to take care to make sure these don't appear in all REST requests for templates since that would cause these to also show up in the site editor.
2. The UI for unlocking blocks may be tricky.  How do we enable this "edit mode" to allow merchants to edit and save templates.
3. If third parties can modify these persisted templates in some way.  I think it's safe at least initially to limit any modification since these would be user defined templates.

**Reply by AnnaMag** [reply]
> If third parties can modify these persisted templates in some way. I think it's safe at least initially to limit any modification since these would be user defined templates.

+1, it makes sense to assume that third parties would modify the standard templates by default, not the user-defined templates.

cc @mattsherman we will explore this direction, so an FYI in case it immediately impacts any of the work on the API/templates.

**Comment by joshuatf** [comment, marked as answer]
We created a template registry and now serve the product templates via REST API as part of https://github.com/woocommerce/woocommerce/pull/39698.  

The solution includes the registry as part of the WC container so it's accessible to all plugins and piggybacks on the existing templates REST endpoint from WordPress core.  If we decide to come up with our own REST endpoint in the future, this should be trivial to swap out the REST service as these services are loosely coupled.

As a follow-up to this, @mattsherman is working on [integrating hooks](https://github.com/woocommerce/woocommerce/discussions/40026) to move template registration and block addition more standardized and to thwart race conditions with block registration.
